blocks/filesystem/base.py

```diff
@@ -21,62 +48,128 @@ def ls(self, path: str) -> Sequence[str]:
         path : str
             The path to the file or directory to list; supports wildcards
         """
-        pass
-
-    @abstractmethod
-    def access(self, paths: Sequence[str]) -> List[DataFile]:
-        """Access multiple paths as file-like objects
-
-        This allows for optimization like parallel downloads
+        protocol, path = self._get_protocol_path(path)
+        fs = self._get_filesystem(protocol)
+        try:
+            if has_magic(path):
+                output = fs.glob(path)
+            else:
+                output = fs.ls(path)
+        # TODO fix in base
+        except FileNotFoundError:
+            return []
+        except NotADirectoryError:
+            return [path]
+
+        if protocol is not None:
+            output = ["://".join([protocol, path]) for path in output]
+        return sorted(output)
+
+    def copy(
+        self,
+        sources: Union[str, Sequence[str]],
+        dest: Union[str, Sequence[str]],
+        recursive=False,
+    ):
+        """Copy the files in sources to dest
 
         Parameters
         ----------
-        paths: list of str
-            The paths of the files to access
-
-        Returns
-        -------
-        files: list of DataFile
-            A list of datafile instances, one for each input path
+        sources : list of str
+            The list of paths to copy
+        dest : str
+            The destination(s) for the copy of source(s)
+        recursive : bool
+            If true, recursively copy any directories
         """
-        pass
+        if isinstance(sources, str):
+            sources = [sources]
+
+        ps, sources = self._get_protocol_path(sources)
+        pd, dest = self._get_protocol_path(dest)
+
+        if ps == pd:
+            fs = self._get_filesystem(ps)
+
+            # Temporary workaround for a bug in gcsfs
+            if ps == "gs" and recursive:
+                sources = fs.expand_path(sources, recursive=True)
+                sources = [s for s in sources if not fs.isdir(s)]
+                return fs.copy(sources, dest, recursive=False)
 
-    @abstractmethod
-    def store(self, bucket: str, files: Sequence[str]) -> ContextManager:
-        """Store multiple data objects
+            fs.copy(sources, dest, recursive=recursive)
 
-        This allows for optimizations when storing several files
+        if ps is None:
+            fs = self._get_filesystem(pd)
+            fs.put(sources, dest, recursive=recursive)
+
+        if pd is None:
+            fs = self._get_filesystem(ps)
+            fs.get(sources, dest, recursive=recursive)
+
+        if pd == "gs":
+            if isinstance(dest, str):
+                fs.invalidate_cache(dest)
```

Comment:
why is this necessary?

Reply:
gcsfs (and probably some of the other implementations) cache many of their operations so future calls to list that destination will potentially not reflect the newly added files. So i am trying to cache invalidate whenever we add new data to the remote filesystem

I'll add an explicit comment here to make this clear to future contributors